File: editor/src/de/project/ice/editor/EntitiesWindow.java
package de.project.ice.editor;

import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.utils.ImmutableArray;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener;
import com.badlogic.gdx.utils.Align;
import com.badlogic.gdx.utils.Array;
import com.kotcrab.vis.ui.util.TableUtils;
import com.kotcrab.vis.ui.widget.VisList;
import com.kotcrab.vis.ui.widget.VisScrollPane;
import com.kotcrab.vis.ui.widget.VisTextButton;
import com.kotcrab.vis.ui.widget.VisWindow;
import de.project.ice.ecs.IceEngine;
import de.project.ice.editor.undoredo.AddEntityAction;
import de.project.ice.editor.undoredo.RemoveEntityAction;
import de.project.ice.editor.undoredo.UndoRedoManager;
import org.jetbrains.annotations.Nullable;


public class EntitiesWindow extends VisWindow
{
    private ImmutableArray<Entity> entities;
    private Array<EntityEntry> currentEntities = new Array<EntityEntry>();
    private VisList<EntityEntry> entityList;
    private EntityEntry selectedEntry = null;
    private SelectionListener selectionListener = null;
    private IceEngine engine;
    private UndoRedoManager undoRedoManager;

    public EntitiesWindow(IceEngine engine, UndoRedoManager undoRedoManager) throws IllegalStateException
    {
        super("Entities");
        this.engine = engine;
        this.entities = engine.getEntities();
        this.undoRedoManager = undoRedoManager;

        TableUtils.setSpacingDefaults(this);
        createWidgets();

        setPosition(0, 0, Align.topRight);
        setResizable(true);
        setSize(200, 400);
    }

    public void setSelectionListener(SelectionListener selectionListener)
    {
        this.selectionListener = selectionListener;
    }

    @Override
    public void act(float delta)
    {
        super.act(delta);
        if (entitiesChanged())
        {
            updateEntities();
        }
        if (selectedEntry != entityList.getSelected())
        {
            selectedEntry = entityList.getSelected();
            if (selectionListener != null)
            {
                selectionListener.selectionChanged(selectedEntry == null ? null : selectedEntry.entity);
            }

        }
    }

    private boolean entitiesChanged()
    {
        // Entities amount changed => we definitely need to update
        if (entities.size() != currentEntities.size)
        {
            return true;
        }

        for (int i = 0; i < entities.size(); ++i)
        {
            if (entities.get(i) != currentEntities.get(i).entity)
            {
                return true;
            }
            else if (!currentEntities.get(i).name.equals(EntityEntry.generateName(entities.get(i))))
            {
                currentEntities.get(i).name = EntityEntry.generateName(entities.get(i));
            }
        }

        return false;
    }

    private void updateEntities()
    {
        currentEntities.clear();

        for (Entity entity : entities)
        {
            currentEntities.add(new EntityEntry(entity));
        }

        entityList.setItems(currentEntities);
    }

    private void createWidgets()
    {

        VisTextButton createEntityBtn = new VisTextButton("Create Entity", new ChangeListener()
        {
            @Override
            public void changed(ChangeEvent event, Actor actor)
            {
                undoRedoManager.addAction(new AddEntityAction(engine.createEntity(), engine));
                updateEntities();
                entityList.setSelectedIndex(entityList.getItems().size-1);
            }
        });
        add(createEntityBtn).expandX().fill();
        VisTextButton deleteEntityBtn = new VisTextButton("Delete selected", new ChangeListener()
        {
            @Override
            public void changed(ChangeEvent event, Actor actor)
            {
                if (entityList.getSelected() != null)
                {
                    undoRedoManager.addAction(new RemoveEntityAction(entityList.getSelected().entity, engine));
                }
            }
        });
        add(deleteEntityBtn).expandX().fill().row();


        entityList = new VisList<EntityEntry>();
        entityList.setWidth(Float.MAX_VALUE);
        updateEntities();

        VisScrollPane scrollPane = new VisScrollPane(entityList);

        row().expand();
        add(scrollPane).colspan(2).fill();
    }

    public interface SelectionListener
    {
        void selectionChanged(@Nullable Entity entity);
    }
}

File: editor/src/de/project/ice/editor/EntitiesWindow.kt
package de.project.ice.editor

import com.badlogic.ashley.core.Entity
import com.badlogic.ashley.utils.ImmutableArray
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener
import com.badlogic.gdx.utils.Align
import com.badlogic.gdx.utils.Array
import com.kotcrab.vis.ui.util.TableUtils
import com.kotcrab.vis.ui.widget.VisList
import com.kotcrab.vis.ui.widget.VisScrollPane
import com.kotcrab.vis.ui.widget.VisTextButton
import com.kotcrab.vis.ui.widget.VisWindow
import de.project.ice.ecs.IceEngine
import de.project.ice.ecs.components.CopyableIceComponent
import de.project.ice.ecs.components.IceComponent
import de.project.ice.editor.undoredo.*


class EntitiesWindow @Throws(IllegalStateException::class)
constructor(private val app: EditorApplication, private val undoRedoManager: UndoRedoManager) : VisWindow("Entities") {
    private val entities: ImmutableArray<Entity> = app.engine.entities
    private val currentEntities = Array<EntityEntry>()
    private var entityList: VisList<EntityEntry>? = null
    private var selectedEntry: EntityEntry? = null
    private var selectionListener: SelectionListener? = null

    init {
        TableUtils.setSpacingDefaults(this)
        createWidgets()

        setPosition(0f, 0f, Align.topRight)
        isResizable = true
        setSize(200f, 400f)
    }

    fun setSelectionListener(selectionListener: SelectionListener) {
        this.selectionListener = selectionListener
    }

    override fun act(delta: Float) {
        super.act(delta)
        if (entitiesChanged()) {
            updateEntities()
        }
        if (selectedEntry !== entityList!!.selected) {
            selectedEntry = entityList!!.selected
            if (selectionListener != null) {
                selectionListener!!.selectionChanged(if (selectedEntry == null) null else selectedEntry!!.entity)
            }

        }
    }

    private fun entitiesChanged(): Boolean {
        // Entities amount changed => we definitely need to update
        if (entities.size() != currentEntities.size) {
            return true
        }

        for (i in 0..entities.size() - 1) {
            if (entities.get(i) !== currentEntities.get(i).entity) {
                return true
            } else if (currentEntities.get(i).name != EntityEntry.generateName(entities.get(i))) {
                currentEntities.get(i).name = EntityEntry.generateName(entities.get(i))
            }
        }

        return false
    }

    private fun updateEntities() {
        currentEntities.clear()

        for (entity in entities) {
            currentEntities.add(EntityEntry(entity))
        }

        entityList!!.setItems(currentEntities)
    }

    private fun createWidgets() {

        val createEntityBtn = VisTextButton("Create", object : ChangeListener() {
            override fun changed(event: ChangeListener.ChangeEvent, actor: Actor) {
                undoRedoManager.addAction(ModifySceneAction.record(app.engine, app.sceneProperties) {
                    it.addEntity(it.createEntity())
                })
                updateEntities()
                entityList!!.selectedIndex = entityList!!.items.size - 1
            }
        })
        add(createEntityBtn).expandX().fill()
        val duplicateEntityBtn = VisTextButton("Duplicate", object : ChangeListener() {
            override fun changed(event: ChangeListener.ChangeEvent, actor: Actor) {
                if (entityList!!.selected != null) {
                    val entity = entityList!!.selected.entity
                    undoRedoManager.addAction(ModifySceneAction.record(app.engine, app.sceneProperties) {
                        val duplicate = it.createEntity()
                        for (component in entity.components) {
                            val iceComponent = component as IceComponent

                            if (iceComponent is CopyableIceComponent) {
                                val dupeComponent = it.createComponent(iceComponent.javaClass)

                                iceComponent.copyTo(dupeComponent)

                                duplicate.add(dupeComponent)
                            }
                        }
                        it.addEntity(duplicate)
                    })
                }
            }
        })
        add(duplicateEntityBtn).expandX().fill()
        val deleteEntityBtn = VisTextButton("Delete", object : ChangeListener() {
            override fun changed(event: ChangeListener.ChangeEvent, actor: Actor) {
                if (entityList!!.selected != null) {
                    val entity = entityList!!.selected.entity
                    undoRedoManager.addAction(ModifySceneAction.record(app.engine, app.sceneProperties) {
                        it.removeEntity(entity)
                    })
                }
            }
        })
        add(deleteEntityBtn).expandX().fill().row()


        entityList = VisList<EntityEntry>()
        entityList!!.width = java.lang.Float.MAX_VALUE
        updateEntities()

        val scrollPane = VisScrollPane(entityList)

        row().expand()
        add(scrollPane).colspan(3).fill()
    }

    interface SelectionListener {
        fun selectionChanged(entity: Entity?)
    }
}